core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -454,102 +421,64 @@ class TiBatchWrite(@transient val df: DataFrame,
+  private def tiKVRow2Key(row: TiRow): SerializableKey =
     new SerializableKey(
-      RowKey.toRowKey(tiTableInfo.getId, extractHandleId(row, isUpdate)).getBytes
+      RowKey.toRowKey(tiTableInfo.getId, extractHandleId(row)).getBytes
     )
 
-  private def sparkRow2TiKVRow(handleId: Long, sparkRow: SparkRow): TiRow = {
+  // convert spark's row to tikv row. We do not allocate handle for no pk case.
+  // allocating handle id will be finished after we check conflict.
+  private def sparkRow2TiKVRow(sparkRow: SparkRow): TiRow = {
     val fieldCount = sparkRow.size
-    var tiRow: TiRow = null
-    if (tiTableInfo.getPrimaryKeyColumn != null) {
-      if (tiTableInfo.isPkHandle) {
-        tiRow = ObjectRowImpl.create(fieldCount)
-        for (i <- 0 until fieldCount) {
-          val data = sparkRow.get(i)
-          val sparkDataType = sparkRow.schema(i).dataType
-          val tiDataType = TiConverter.fromSparkType(sparkDataType)
-          tiRow.set(i, tiDataType, data)
-        }
-      } else if (tiTableInfo.getAutoIncrementColInfo != null) {
-        // when column is auto_increment, it must has a primary key
-        // its value will be filled at the beginning according tidb's logic.
-        val autoincrementCol = tiTableInfo.getAutoIncrementColInfo
-        tiRow = ObjectRowImpl.create(fieldCount)
-        for (i <- 0 until fieldCount) {
-          var data = sparkRow.get(i)
-          val sparkDataType = sparkRow.schema(i).dataType
-          val colName = sparkRow.schema(i).name
-          // check do we need fill auto increment column
-          if (colName.equals(autoincrementCol.getName)) {
-            // only update auto increment column when the value is null
-            if (data == null) {
-              data = handleId
-            }
-          }
-          val tiDataType = TiConverter.fromSparkType(sparkDataType)
-          tiRow.set(i, tiDataType, data)
-        }
-      } else {
-        throw new TiBatchWriteException(
-          "cannot insert on a table with primary key but it's not a handle" +
-            "or auto increment"
-        )
-      }
-    } else {
-      // table does not have primary key, we can just insert and do not need consider
-      // the constraint of primary key.
-      tiRow = ObjectRowImpl.create(fieldCount + 1)
-      for (i <- 0 until fieldCount) {
-        val data = sparkRow.get(i)
-        val sparkDataType = sparkRow.schema(i).dataType
-        //val tiDataType = TiConverter.fromSparkType(sparkDataType)
-        tiRow.set(i, null, data)
-      }
-      // append _tidb_rowid at the end
-      tiRow.set(fieldCount, IntegerType.BIGINT, handleId)
+    val tiRow = ObjectRowImpl.create(fieldCount)
+    for (i <- 0 until fieldCount) {
+      val data = sparkRow.get(i)
+      val sparkDataType = sparkRow.schema(i).dataType
+      val tiDataType = TiConverter.fromSparkType(sparkDataType)
+      tiRow.set(i, tiDataType, data)
```

Comment:
`tiDataType` is ignored in `ObjectRowImpl`, remove ` val tiDataType = TiConverter.fromSparkType(sparkDataType)` to let ci pass

Reply:
done.